Initial support DipTrace 4.3 gerber files.

hydra opened the pull request:
I've been using [DipTrace ](https://diptrace.com/)for many years, however it seems to export non-standard/off-spec gerber files.

Issues:
* Diptrace is not using a leading `0` on the `D0*` commands, for example it writes `D1` instead of `D01`.

PR status:

Not ready for merge yet, I'm in currently working with various gerber files and there may be more issues to resolve. As I find them I'll add them to this PR as individual commits.

hydra commented:
Other errors on a test file include:

```
[2025-04-23T08:27:54Z WARN  gerber_parser::parser] Error: Document included a line that isn't supported.
    Line: '7: %AMOUTLINE0*'
[2025-04-23T08:27:54Z WARN  gerber_parser::parser] Error: Document included a line that isn't valid.
    Line: '8: 4,1,4,'
[2025-04-23T08:27:54Z WARN  gerber_parser::parser] Error: Document included a line that isn't valid.
    Line: '9: -0.24999,0.27501,'
```

```
[2025-04-23T08:27:55Z WARN  gerber_parser::parser] Error: The Aperture attribute '.AperFunction,ComponentPad*%' is not supported, but presumably valid.
    Line: '388: %TA.AperFunction,ComponentPad*%'
```

```
[2025-04-23T08:27:55Z WARN  gerber_parser::parser] Error: The Aperture attribute '.AperFunction,ViaPad*%' is not supported, but presumably valid.
    Line: '397: %TA.AperFunction,ViaPad*%'
````

```
[2025-04-23T08:27:55Z WARN  gerber_parser::parser] Error: Command was uniquely identified, but did not match regex: %ADD([0-9]+)([A-Z]),(.*)\*%.
    Line: '399: %ADD100OUTLINE0*%'
```

```
[2025-04-23T08:27:56Z WARN  gerber_parser::parser] Error: tried to parse the selection of aperture '102' but it is not defined.
    Line: '5728: D102*'
```

The file I'm using is this one:

[1 - L1_TOP_SIGNAL.zip](https://github.com/user-attachments/files/19864079/1.-.L1_TOP_SIGNAL.zip)

I've not investigated the above warnings yet, but if you have any suggestions please let me know.

hydra commented:
Added support for parsing 'Outline' aperture macros.

Example project and file used for testing:
[rotated_ic.zip](https://github.com/user-attachments/files/19895004/rotated_ic.zip)

screenshot of it in diptrace's gerber viewer:

![image](https://github.com/user-attachments/assets/8666079c-95f6-4e2f-ac1f-0eea469e9276)

hydra commented:
Some background: I'm working on a rust-based Gerber viewer, for my MakerPnP project.

https://github.com/MakerPnP/makerpnp/tree/gerber-viewer/crates/gerber_viewer_egui

here's a video of it as it is today:

https://github.com/user-attachments/assets/fae467b4-e15d-4ece-8bf9-cc47c645b3c1

hydra commented:
here's my renderer parsing the rotated_ic.zip example, which has:

* Diptrace style short `Dx` commands.
* Some commands with missing X/Y coordinates.  Note that the renderer maintains the X/Y coordinate, not the parser.
* Aperture macros.

![image](https://github.com/user-attachments/assets/8ea344e2-8c1d-4216-ae03-28bff3b1292e)

NemoAndrea commented:
A nice set of improvement to the capabilities of the crate; thank you for contributing! I don't know enough about Gerber files to know if supporting the off-spec [DipTrace](https://diptrace.com/) format would cause problems for other files, but that will be a problem for another day.

I can merge and bump the crate version once I understand issue https://github.com/NemoAndrea/gerber-parser/issues/3

hydra commented:
Hi, added a comment on #3.  I've also squashed a couple of commits in this branch.

I've also opened #5 which has some much needed support for macros and macro expressions which builds on this PR and supports KiCad 8.

It might be easier to just review the final code of #5, since the commits in this PR are used in #5, but keep the individual commits in mind as you review the changes on both PRs.

With regards to potential breakages, I don't think there are any, since all the tests still pass and new ones have been added in the this PR and also #5.

hydra commented:
@NemoAndrea See https://github.com/dbrgn/gerber-types-rs/pull/36#issuecomment-2843350520

I've got my renderer rendering Aperture Macros.

Here's a debug log from my app as it builds the primitives from this test gerber file:

```
%TF.GenerationSoftware,KiCad,Pcbnew,8.0.3*%
%TF.CreationDate,2025-04-28T16:25:44+02:00*%
%TF.SameCoordinates,Original*%
%TF.FileFunction,Copper,L4,Bot*%
%TF.FilePolarity,Positive*%
%FSLAX46Y46*%
G04 Gerber Fmt 4.6, Leading zero omitted, Abs format (unit mm)*
G04 Created by KiCad (PCBNEW 8.0.3) date 2025-04-28 16:25:44*
%MOMM*%
%LPD*%
G01*
G04 APERTURE LIST*
G04 Aperture macros list*
%AMRoundRect*
0 Rectangle with rounded corners*
0 $1 Rounding radius*
0 $2 $3 $4 $5 $6 $7 $8 $9 X,Y pos of 4 corners*
0 Add a 4 corners polygon primitive as box body*
4,1,4,$2,$3,$4,$5,$6,$7,$8,$9,$2,$3,0*
0 Add four circle primitives for the rounded corners*
1,1,$1+$1,$2,$3*
1,1,$1+$1,$4,$5*
1,1,$1+$1,$6,$7*
1,1,$1+$1,$8,$9*
0 Add four rect primitives between the rounded corners*
20,1,$1+$1,$2,$3,$4,$5,0*
20,1,$1+$1,$4,$5,$6,$7,0*
20,1,$1+$1,$6,$7,$8,$9,0*
20,1,$1+$1,$8,$9,$2,$3,0*%
%ADD39RoundRect,X0.500000X0.000000X0.000000X5.000000X0.000000X5.000000X5.000000X0.000000X5.000000*%
D39*
X005000000Y-005000000D03*
M02*

```

```
[2025-05-01T06:35:56Z WARN  gerber_parser::parser] Error: The AttributeName 'GenerationSoftware' is currently not supported for File Attributes.
    Line: '1: %TF.GenerationSoftware,KiCad,Pcbnew,8.0.3*%'
[2025-05-01T06:35:56Z WARN  gerber_parser::parser] Error: The AttributeName 'CreationDate' is currently not supported for File Attributes.
    Line: '2: %TF.CreationDate,2025-04-28T16:25:44+02:00*%'
[2025-05-01T06:35:56Z WARN  gerber_parser::parser] Error: The AttributeName 'SameCoordinates' is currently not supported for File Attributes.
    Line: '3: %TF.SameCoordinates,Original*%'
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(FileAttribute(FileFunction(Other("Copper"))))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(FileAttribute(FilePolarity(Positive)))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(CoordinateFormat(CoordinateFormat { integer: 4, decimal: 6 }))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(GCode(Comment("Gerber Fmt 4.6, Leading zero omitted, Abs format (unit mm)")))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(GCode(Comment("Created by KiCad (PCBNEW 8.0.3) date 2025-04-28 16:25:44")))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(Unit(Millimeters))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(LoadPolarity(Dark))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(GCode(InterpolationMode(Linear)))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(GCode(Comment("APERTURE LIST")))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(GCode(Comment("Aperture macros list")))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(ApertureMacro(ApertureMacro { name: "RoundRect", content: [Comment("Rectangle with rounded corners"), Comment("$1 Rounding radius"), Comment("$2 $3 $4 $5 $6 $7 $8 $9 X,Y pos of 4 corners"), Comment("Add a 4 corners polygon primitive as box body"), Outline(OutlinePrimitive { e
xposure: Value(true), points: [(Variable(2), Variable(3)), (Variable(4), Variable(5)), (Variable(6), Variable(7)), (Variable(8), Variable(9)), (Variable(2), Variable(3))], angle: Value(0.0) }), Comment("Add four circle primitives for the rounded corners"), Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(2
), Variable(3)), angle: None }), Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(4), Variable(5)), angle: None }), Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(6), Variable(7)), angle: None }), Circle(CirclePrimitive { exposure: Value(true), diameter: Ex
pression("$1+$1"), center: (Variable(8), Variable(9)), angle: None }), Comment("Add four rect primitives between the rounded corners"), VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(2), Variable(3)), end: (Variable(4), Variable(5)), angle: Value(0.0) }), VectorLine(VectorLinePrimitive { exposure: Va
lue(true), width: Expression("$1+$1"), start: (Variable(4), Variable(5)), end: (Variable(6), Variable(7)), angle: Value(0.0) }), VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(6), Variable(7)), end: (Variable(8), Variable(9)), angle: Value(0.0) }), VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(8), Variable(9)), end: (Variable(2), Variable(3)), angle: Value(0.0) })] }))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: ExtendedCode(ApertureDefinition(ApertureDefinition { code: 39, aperture: Macro("RoundRect", Some([Value(0.5), Value(0.0), Value(0.0), Value(5.0), Value(0.0), Value(5.0), Value(5.0), Value(0.0), Value(5.0)])) }))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(DCode(SelectAperture(39)))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(DCode(Operation(Flash(Coordinates { x: Some(CoordinateNumber { nano: 5000000 }), y: Some(CoordinateNumber { nano: -5000000 }), format: CoordinateFormat { integer: 4, decimal: 6 } }))))
[2025-05-01T06:35:56Z DEBUG gerber_parser::parser] Found command: FunctionCode(MCode(EndOfFile))
[2025-05-01T06:35:56Z INFO  gerber_viewer_egui] Gerber file parsed successfully. path: D:\Users\Hydra\Documents\dev\projects\makerpnp\makerpnp\crates\gerber_viewer_egui\examples\kicad_8.0\macros_1\macros_1.gbr
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] macro_name_and_args: RoundRect_0.5X0X0X5X0X5X5X0X5
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] macro_context: MacroContext { variables: {4: 5.0, 2: 0.0, 3: 0.0, 6: 5.0, 1: 0.5, 5: 0.0, 7: 5.0, 9: 5.0, 8: 0.0} }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("Rectangle with rounded corners")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("$1 Rounding radius")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("$2 $3 $4 $5 $6 $7 $8 $9 X,Y pos of 4 corners")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("Add a 4 corners polygon primitive as box body")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Outline(OutlinePrimitive { exposure: Value(true), points: [(Variable(2), Variable(3)), (Variable(4), Variable(5)), (Variable(6), Variable(7)), (Variable(8), Variable(9)), (Variable(2), Variable(3))], angle: Value(0.0) })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("Add four circle primitives for the rounded corners")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(2), Variable(3)), angle: None })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(4), Variable(5)), angle: None })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(6), Variable(7)), angle: None })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Circle(CirclePrimitive { exposure: Value(true), diameter: Expression("$1+$1"), center: (Variable(8), Variable(9)), angle: None })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: Comment("Add four rect primitives between the rounded corners")
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(2), Variable(3)), end: (Variable(4), Variable(5)), angle: Value(0.0) })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(4), Variable(5)), end: (Variable(6), Variable(7)), angle: Value(0.0) })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(6), Variable(7)), end: (Variable(8), Variable(9)), angle: Value(0.0) })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] content: VectorLine(VectorLinePrimitive { exposure: Value(true), width: Expression("$1+$1"), start: (Variable(8), Variable(9)), end: (Variable(2), Variable(3)), angle: Value(0.0) })
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] primitive_defs: [Polygon { center: Position { x: 0.0, y: 0.0 }, vertices: [Position { x: 0.0, y: 0.0 }, Position { x: 5.0, y: 0.0 }, Position { x: 5.0, y: 5.0 }, Position { x: 0.0, y: 5.0 }, Position { x: 0.0, y: 0.0 }], exposure: Add, is_convex: true, triangles: [] }, Circle { center: Position { x: 0.0, y
: 0.0 }, diameter: 1.0, exposure: Add }, Circle { center: Position { x: 5.0, y: 0.0 }, diameter: 1.0, exposure: Add }, Circle { center: Position { x: 5.0, y: 5.0 }, diameter: 1.0, exposure: Add }, Circle { center: Position { x: 0.0, y: 5.0 }, diameter: 1.0, exposure: Add }, Polygon { center: Position { x: 2.5, y: 0.0 }, vertices: [Position { x: -2.5, y:
 -0.5 }, Position { x: -2.5, y: 0.5 }, Position { x: 2.5, y: 0.5 }, Position { x: 2.5, y: -0.5 }], exposure: Add, is_convex: true, triangles: [] }, Polygon { center: Position { x: 5.0, y: 2.5 }, vertices: [Position { x: 0.5, y: -2.5 }, Position { x: -0.5, y: -2.5 }, Position { x: -0.5, y: 2.5 }, Position { x: 0.5, y: 2.5 }], exposure: Add, is_convex: tr
ue, triangles: [] }, Polygon { center: Position { x: 2.5, y: 5.0 }, vertices: [Position { x: 2.5, y: 0.5 }, Position { x: 2.5, y: -0.5 }, Position { x: -2.5, y: -0.5 }, Position { x: -2.5, y: 0.5 }], exposure: Add, is_convex: true, triangles: [] }, Polygon { center: Position { x: 0.0, y: 2.5 }, vertices: [Position { x: -0.5, y: 2.5 }, Position { x: 0.5, y: 2.5 }, Position { x: 0.5, y: -2.5 }, Position { x: -0.5, y: -2.5 }], exposure: Add, is_convex: true, triangles: [] }]
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Polygon { center: Position { x: 5.0, y: -5.0 }, vertices: [Position { x: 0.0, y: 0.0 }, Position { x: 5.0, y: 0.0 }, Position { x: 5.0, y: 5.0 }, Position { x: 0.0, y: 5.0 }, Position { x: 0.0, y: 0.0 }], exposure: Add, is_convex: true, triangles: [] }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Circle { center: Position { x: 5.0, y: -5.0 }, diameter: 1.0, exposure: Add }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Circle { center: Position { x: 10.0, y: -5.0 }, diameter: 1.0, exposure: Add }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Circle { center: Position { x: 10.0, y: 0.0 }, diameter: 1.0, exposure: Add }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Circle { center: Position { x: 5.0, y: 0.0 }, diameter: 1.0, exposure: Add }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Polygon { center: Position { x: 7.5, y: -5.0 }, vertices: [Position { x: -2.5, y: -0.5 }, Position { x: -2.5, y: 0.5 }, Position { x: 2.5, y: 0.5 }, Position { x: 2.5, y: -0.5 }], exposure: Add, is_convex: true, triangles: [] }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Polygon { center: Position { x: 10.0, y: -2.5 }, vertices: [Position { x: 0.5, y: -2.5 }, Position { x: -0.5, y: -2.5 }, Position { x: -0.5, y: 2.5 }, Position { x: 0.5, y: 2.5 }], exposure: Add, is_convex: true, triangles: [] }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Polygon { center: Position { x: 7.5, y: 0.0 }, vertices: [Position { x: 2.5, y: 0.5 }, Position { x: 2.5, y: -0.5 }, Position { x: -2.5, y: -0.5 }, Position { x: -2.5, y: 0.5 }], exposure: Add, is_convex: true, triangles: [] }
[2025-05-01T06:35:56Z DEBUG gerber_viewer_egui] flashing macro primitive: Polygon { center: Position { x: 5.0, y: -2.5 }, vertices: [Position { x: -0.5, y: 2.5 }, Position { x: 0.5, y: 2.5 }, Position { x: 0.5, y: -2.5 }, Position { x: -0.5, y: -2.5 }], exposure: Add, is_convex: true, triangles: [] }
[2025-05-01T06:35:59Z DEBUG gerber_viewer_egui] view bbox: BoundingBox { min_x: 4.5, min_y: -5.5, max_x: 10.5, max_y: 0.5 }
```

Of note, the macro has variables (e.g. '$1') and expressions ('$1+$1'), which my renderer handles and evaluates.

Here's a screenshot with "Unique shape colors" enabled, so it's easy to see the results.

![image](https://github.com/user-attachments/assets/9b6a2563-2e6b-45b3-a09c-94df12c68efa)
